orttraining/orttraining/eager/ort_aten.cpp

```diff
@@ -920,6 +920,55 @@ const at::Tensor& resize_(
   return self;
 }
 
+// aten::cat.out(Tensor[] tensors, int dim=0, *, Tensor(a!) out) -> Tensor(a!)
+at::Tensor& cat_out(
+  at::TensorList tensors,
+  int64_t dim,
+  // *,
+  at::Tensor& out) {
+  ORT_LOG_FN(tensors, dim, out);
+
+  assert(tensors.size() > 0);
+  if (
+      std::vector<at::ScalarType> supportedTypes =
+        {at::kBFloat16, at::kBool, at::kByte, at::kDouble, at::kFloat, at::kHalf, at::kInt, at::kLong, at::kShort};
+      !IsSupportedType(tensors, supportedTypes)) {
+    return at::native::call_fallback_fn<
+      &at::native::cpu_fallback,
+      ATEN_OP(cat_out)>::call(tensors, dim, out);
+  }
+  auto& invoker = GetORTInvoker(tensors[0].device());
```

Comment:
I think the above assert on size of tensor is only run in debug builds - do we need to perform bounds checking for non-debug builds as well?

Reply:
if size is zero then it'll crash on deference. it would be unexpected to concat 0 tensors together

Reply:
I was wondering if it was preferrable to throw some runtime error vs crash on dereference (or if there is some existing pattern for how this is handled in general)... I guess a similar question could be asked if the shape of the tensors is validated  before this operation. If the concat dimension is larger then the number of dimensions in one of the tensors...

Reply:
I am fine with the current implementation - feel free to close / resolve this.

Reply:
it kind-of depends.  If you call cat without an out, it'll check/normalize everything before calling out.  But if you give it an out, then it goes directly to cat.out without the checks